src/state/export/exportWhiteboard.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2023 Nordeck IT + Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import log from 'loglevel';
+import { convertBlobToBase64, convertMxcToHttpUrl, isDefined } from '../../lib';
+import {
+  Element,
+  ImageElement,
+  WhiteboardDocument,
+  getElement,
+  getNormalizedElementIds,
+  getNormalizedSlideIds,
+  getSlideLock,
+} from '../crdt';
+import { SharedMap } from '../crdt/y';
+import {
+  WhiteboardDocumentExport,
+  WhiteboardFileExport,
+} from './whiteboardDocumentExport';
+
+/**
+ * Export a whiteboard including files.
+ * If a file download raises an error the data won't be included into the export.
+ * However the information about the image element itself will still be part of the whiteboard data.
+ *
+ * @param doc - Document of the whiteboard to export
+ * @param baseUrl - Homeserver base URL to download the files
+ *
+ * @returns A Promise that resolves to a whiteboard export
+ */
+export async function exportWhiteboard(
+  doc: SharedMap<WhiteboardDocument>,
+  baseUrl: string,
+): Promise<WhiteboardDocumentExport> {
+  return {
+    version: 'net.nordeck.whiteboard@v1',
+    whiteboard: {
+      slides: getNormalizedSlideIds(doc).map((slideId) => ({
+        elements: getNormalizedElementIds(doc, slideId)
+          .map((elementId) => getElement(doc, slideId, elementId)?.toJSON())
+          .filter(isDefined),
+        lock: getSlideLock(doc, slideId) ? {} : undefined,
+      })),
+      files: await exportFiles(doc, baseUrl),
```

Comment:
Should we care that if image id duplicated we could have a one entry in `files`?

Reply:
Good point. There should be a deduplication in the export data.